cpp/src/interop/from_arrow_host.cu

```diff
@@ -0,0 +1,445 @@
+                                     input->null_count,
+                                     std::move(*get_mask_buffer(input)));
+
+  return input->offset == 0
+           ? std::move(out_col)
+           : std::make_unique<column>(
+               cudf::detail::slice(out_col->view(),
+                                   static_cast<size_type>(input->offset),
+                                   static_cast<size_type>(input->offset + input->length),
+                                   stream),
+               stream,
+               mr);
+}
+
+template <>
+std::unique_ptr<column> dispatch_copy_from_arrow_host::operator()<cudf::dictionary32>(
+  ArrowSchemaView* schema, ArrowArray const* input, data_type type, bool skip_mask)
+{
+  ArrowSchemaView keys_schema_view;
+  NANOARROW_THROW_NOT_OK(
+    ArrowSchemaViewInit(&keys_schema_view, schema->schema->dictionary, nullptr));
+
+  auto const keys_type = arrow_to_cudf_type(&keys_schema_view);
+  auto keys_column =
+    get_column_copy(&keys_schema_view, input->dictionary, keys_type, true, stream, mr);
+
+  auto const dict_indices_type = [&schema]() -> data_type {
+    // cudf dictionary requires an unsigned type for the indices,
+    // since it is invalid for an arrow dictionary to contain negative
+    // indices, we can safely use the unsigned equivalent without having
+    // to modify the buffers.
+    switch (schema->storage_type) {
+      case NANOARROW_TYPE_INT8:
+      case NANOARROW_TYPE_UINT8: return data_type(type_id::UINT8);
+      case NANOARROW_TYPE_INT16:
+      case NANOARROW_TYPE_UINT16: return data_type(type_id::UINT16);
+      case NANOARROW_TYPE_INT32:
+      case NANOARROW_TYPE_UINT32: return data_type(type_id::UINT32);
+      case NANOARROW_TYPE_INT64:
+      case NANOARROW_TYPE_UINT64: return data_type(type_id::UINT64);
+      default: CUDF_FAIL("Unsupported type_id for dictionary indices", cudf::data_type_error);
+    }
+  }();
+
+  auto indices_column = get_column_copy(schema, input, dict_indices_type, false, stream, mr);
+  // child columns shouldn't have masks and we need the mask in the main column
+  auto column_contents = indices_column->release();
+  indices_column       = std::make_unique<column>(dict_indices_type,
+                                            static_cast<size_type>(input->length),
+                                            std::move(*(column_contents.data)),
+                                            rmm::device_buffer{},
+                                            0);
+
+  return make_dictionary_column(std::move(keys_column),
+                                std::move(indices_column),
+                                std::move(*(column_contents.null_mask)),
+                                input->null_count);
+}
+
+template <>
+std::unique_ptr<column> dispatch_copy_from_arrow_host::operator()<cudf::struct_view>(
+  ArrowSchemaView* schema, ArrowArray const* input, data_type type, bool skip_mask)
+{
+  char buffer[1024];
```

Comment:
Is this used anywhere? I see it being passed to the lambda, but it's not accessed there.